### PR TITLE
Feature - shadow options

### LINKED ISSUE #18: [Enchancement] Enhance shadow handling
### **Is your enhancement request related to a problem? Please describe.**
The current implementation of shadows doesn't handle very well. It draws a border but doesn't fill to provide a nice shadow, so blur doesn't impact much when you adjust the value.

### **Describe the enhancement you'd like**
I would like the feature to improve shadow handling, to both improve the visual appearance of shadows and to increase the amount of control a user has on a shadow.

### **Describe alternatives you've considered & Possible Implementation Details**
I've already built a solution for this, which provides a small `popover` menu when a settings button is clicked, providing the user with more options around shadows. Other options include `X offset`, `Y offset`, `color` and `blur` is also included (but displays any existing state provided by the slider).

This improvement also improves the visuals of shadows on the UI, making it a much nicer UX than the existing implementation which only provides a border/frame shadow. This new suggestion also maps transparent images to provide an accurate image shadow. If border radius is added to a `png` with a transparent background then it provides a filled background based on `shadow.color` and rounds the edges of the fill. Further context provided below.

### **Additional context**
I'd like to open a PR with the following implementation if you're happy for me to:

**Existing**
Screen Recording 2024-09-30 at 12 44 10

**Suggested implementation**
Screen Recording 2024-09-30 at 12 43 08

---

### **Optional Sections (if relevant):**

- **Priority (Low / Medium / High)**  Low
- **Are you willing to submit a PR for this enhancement?**  Yes
- **Does this enhancement require changes in documentation?** No

### PR DESCRIPTION
**Pull request by kvtemadden**
### **Description**
Provides the user with more options/control over image shadows. 

### **Related Issue**
[18](https://github.com/suryanshsingh2001/mockly/issues/18)

Closes #18 

### **Changes Made**
- Used `roundRect` for shadow to provide filled shadow
- Implemented shadow to match image shape if `borderRadius` is not set
- Added `popover` component and created `shadowManager` to display options on UI

### **Type of Change**
_Select the type of change made with this pull request:_

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### **Screenshots (if applicable)**
Screen Recording 2024-09-30 at 12 43 08

### **How Has This Been Tested?**
- _Manually tested on local environment_

### **Checklist**
_Check off the following items before submitting the pull request:_

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by suryanshsingh2001** [comment]
@kvtemadden Hey, please take a pull for the latest changes from `master` branch. And add these changes after it.

**Comment by kvtemadden** [comment]
Hey @suryanshsingh2001, just merged main into this branch & resolved conflicts. 

> And add these changes after it

Was there something you wanted me to update?

**Comment by suryanshsingh2001** [comment]
> Hey @suryanshsingh2001, just merged main into this branch & resolved conflicts.
> 
> > And add these changes after it
> 
> Was there something you wanted me to update?

No, just the contents of this pull request.

**Comment by suryanshsingh2001** [comment]
Hi @kvtemadden,

I’m experiencing a potential memory leak or performance issue with the browser, as the RAM usage is abnormally high. I’ve captured a screen recording that illustrates the problem in preview mode for vercel.

Please take a look for any potential memory leaks or performance overhead. I have tested in the latest version so far no issues.

https://github.com/user-attachments/assets/ff25a5ed-aa30-4c1d-81b7-8b1804304f4a

**Comment by kvtemadden** [comment]
Hey @suryanshsingh2001, I think this might've been to do with how the image save, use and paint to canvas was working since fixing the canvas flickering.

I've gone through and tweaked a few parts to make it more reliable, I can't repro the issue but could you give it another go please?

File: components/shared/Editor.tsx [hunk 1]
```diff
@@ -315,8 +319,6 @@ export default function MockupEditor() {
 
       drawBackgroundImage(ctx);
     }
-
-    requestAnimationFrame(drawCanvas);
```
Comment (suryanshsingh2001) [inline components/shared/Editor.tsx:316]: Hey @kvtemadden, is requestAnimation not required anymore>

**Comment by suryanshsingh2001** [comment]
Hi @kvtemadden,

It seems the issue isn't occurring anymore. I suspect it might have been a bug with the Vercel preview environment, as I wasn't able to reproduce it in my local setup. It looks like the issue appeared unexpectedly and has since resolved itself.

File: components/shared/Editor.tsx (hunk 1 above)
Comment (kvtemadden) [reply components/shared/Editor.tsx:316]: It seems not, I believe handling the image loading more efficiently resolved the situation without needing to use `requestAnimationFrame`. I had originally tried that route on the initial fix as many people suggested this as the resolution when I was researching, but hadn't got it quite right. It seems fresh eyes resolved the problem & saved using it unnecessarily.

Comment (suryanshsingh2001) [reply components/shared/Editor.tsx:316]: Okay, this is fantastic! I can definitely see a reduction in RAM usage on the preview. It looks like the adjustments you made to handle image loading more efficiently have really helped.

Great work!

**Review by suryanshsingh2001 — APPROVED** [review]

**Comment by suryanshsingh2001** [comment]
Hi @kvtemadden,

I wanted to take a moment to personally thank you for your valuable contributions. Your insights and solutions have truly shaped the way Mockly has progressed, and I’m incredibly grateful for your work.

If there are any other issues you’d like to explore or areas for further improvement, feel free to dive in. I’d love the opportunity to collaborate with you again in the future.

Thanks once again, and I look forward to working with you soon!

**Pull request merged**